### PR TITLE
feedback_final

### PR DESCRIPTION
**Pull request by irrso**

### CODE REVIEW

File: feat_eng/FE_gw.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+# In[1]:
+
+
+import numpy as np
+import pandas as pd
+import os
+import gc
+import utils
+import feat_eng
+# import dr_clust
```
Comment (sally-yeom) [inline feat_eng/FE_gw.py:15]: 사용하지 않는 import 주석은 제거하시는게 좋아보입니다.

File: feat_eng/FE_gw.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+
+df = feat_eng.create_complex_id(df=df, lat_col='latitude', lon_col='longitude')
+
+
+# In[8]:
+
+
+subwayInfo = pd.read_csv(os.path.join(file_path, 'subwayInfo.csv'))
+schoolInfo = pd.read_csv(os.path.join(file_path, 'schoolinfo.csv'))
+parkInfo = pd.read_csv(os.path.join(file_path, 'parkInfo.csv'))
```
Comment (sally-yeom) [inline feat_eng/FE_gw.py:73]: 해당 내용은 변수명 형식이 Camel Case인 경우인데요, 몇 줄 위의 코드를 보면 Snake Case도 혼재되어 사용되고 있습니다. 

`dup_train = train.drop('index', axis=1).duplicated()`

정책에 따라 다르겠으나, 보통 파이썬에서는 변수명을 Snake Case로 사용합니다.
또한 pandas dataframe이라는 점을 알 수 있도록 df 로 변수명이 끝나면 더 가독성이 좋아질 것 같습니다.

File: feat_eng/FE_gw.py [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+# In[12]:
+
+
+interestRate = pd.read_csv(os.path.join(file_path, 'interestRate.csv')).rename({'year_month': 'contract_year_month'}, axis=1)
+
+
+# In[13]:
+
+
+df = df.merge(feat_eng.add_interest_rate(interestRate), on='contract_year_month', how='left')
```
Comment (sally-yeom) [inline feat_eng/FE_gw.py:144]: 단순히 df가 아닌 어떤 정보를 담고 있는 df 인 것을 변수명에 명시해주면 좋겠습니다.

File: feat_eng/FE_gw.py [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+del monthly_new_supply
+
+
+# # 5. 클러스터 변수 붙여주기
+# 
+# - `Standard_None_KMenas100_results.csv`:
+#     - `['latittude', 'longitude']`
+# - `Minmax_UMAP02_KMeans500_results.csv`:
+#     - `['latitude', 'longitude', 'nearest_subway_distance_km', '2023_age', 'deposit_mean', 'deposit_std']`
```
Comment (sally-yeom) [inline feat_eng/FE_gw.py:166]: 최종 제출 코드에 불필요한 cell(ipynb을 py로 옮기신 것이므로)은 삭제해주시는게 좋습니다.

File: feat_eng/feat_eng.py [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+    unique_complexes['complex_id'] = unique_complexes.index
+
+    # 원본 df에 'complex_id' 추가
+    df_with_id = df.merge(unique_complexes, on=[lat_col, lon_col], how='left')
+    
+    return df_with_id
+
+# Distance to the Nearest POI (km)
+def nearest_POI(apt, POI):
+    from sklearn.neighbors import NearestNeighbors
```
Comment (sally-yeom) [inline feat_eng/feat_eng.py:30]: 코드의 가독성을 위해 nested import 보다는 파일 상단에 import를 모아두시는게 좋을 것 같습니다.
해당 함수 내에 import 해서 런타임에서 얻는 성능적인 이점은 별로 크지 않습니다.

File: feat_eng/feat_eng.py [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+    
+    return result_df
+
+# 층수 및 면적 범주화
+# 층수 범주화 함수
+def categorize_floor(df):
+    """
+    각 아파트 단지(complex_id) 내에서 층수를 저층, 중층, 고층으로 범주화합니다.
+    각 단지별로 최대 층수를 기준으로 저층, 중층, 고층을 자동으로 결정합니다.
+    """
```
Comment (sally-yeom) [inline feat_eng/feat_eng.py:295]: 해당 파일에는 여러개의 함수가 존재하고 다음과 같은 경우가 다 존재하는데요.

```
- 함수명 상단의 주석으로 간단히 설명한 경우
- 함수명 아래에 docstring으로 설명한 경우
- 하나의 함수에 두가지가 다 존재
```

해당 주석 형식을 한가지로 통일해주시면 좋을 것 같습니다. docstring이 더 명시적이긴 할 것입니다.

File: feat_eng/feat_eng2.py [hunk 7]
```diff
@@ -0,0 +1,496 @@
+import pandas as pd
+import numpy as np
+
+# 파일 경로 및 파일 불러오기
+data = pd.read_csv('/data/ephemeral/home/level2-competitiveds-recsys-05/notebooks/JWORGINIZETEST/train_label.csv')
```
Comment (sally-yeom) [inline feat_eng/feat_eng2.py:5]: 이 파일 의외에도 많은 파일에서 read_csv, to_csv 함수 명의 경로가 해당 경우처럼 string으로 들어가 있는데요.
경로 및 상수같은 경우는 나중에 config 파일을 만들어 별도로 경로를 관리하는게 좋습니다.

File: feat_eng/feat_eng2.py [hunk 8]
```diff
@@ -0,0 +1,496 @@
...
+data = data[data['_type'] == 'train']
+# data = data[data['contract_year_month'] == 202303]
```
Comment (sally-yeom) [inline feat_eng/feat_eng2.py:7]: 사용하지 않는 코드는 최종 제출 시에는 삭제해야 합니다.

File: feat_eng/feat_eng2.py [hunk 9]
```diff
@@ -0,0 +1,496 @@
...
+
+data2 = pd.read_csv('/data/ephemeral/home/level2-competitiveds-recsys-05/notebooks/EDA&FE/final.csv')
+data['cluster_labels2'] = labels
+cluster_data = data[['cluster_labels2', 'complex_id']]
+
+data3 = data2.merge(cluster_data, on=['complex_id'], how='left')
+data3.to_csv('final2.csv', index = False)
+
+import pandas as pd
+from prophet import Prophet
```
Comment (sally-yeom) [inline feat_eng/feat_eng2.py:79]: import 구문은 파일 상단에 모아서 위치시켜야 합니다.

File: feat_eng/utils.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+
+def reduce_mem_usage(df, exclude_columns=None):
+    """
+    Iterate through all the columns of a dataframe and modify the data type
+    to reduce memory usage while maintaining numeric precision.
+    Parameters:
+    df (DataFrame): The dataframe to optimize.
+    exclude_columns (list): List of column names to exclude from type conversion.
+    """
```
Comment (sally-yeom) [inline feat_eng/utils.py:12]: 해당 파일은 모든 주석이 영어로 되어있습니다.
다른 파일의 주석과 언어를 통일해주시면 좋을 것 같습니다.

File: modeling/CatBoost.py [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+
+RANDOM_SEED = 42
+np.random.seed(RANDOM_SEED)
+
+# 파일 경로 설정
+file_path = '../data/'
+
+# 파일 불러오기
+df = pd.read_csv(file_path + 'final.csv')
+sample_submission = pd.read_csv(file_path + 'sample_submission.csv')
```
Comment (sally-yeom) [inline modeling/CatBoost.py:24]: RANDOM_SEED로 선언되는 상수값과 파일 경로 및 파일 명 들은 config 파일을 만들어 별도로 관리하시는게 가독성 및 관리 면에서 좋습니다.

[yaml 파일로 생성](https://velog.io/@bobaebak/Python-%EC%84%A4%EC%A0%95-%ED%8C%8C%EC%9D%BC-YAML-%EC%82%AC%EC%9A%A9%ED%95%98%EA%B8%B0)할 수 있고, config 역할을 하는 py 파일을 만들 수도 있습니다.

File: modeling/CatBoost.py [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+       'deposit_label', 'nearest_subway_distance_km',
+       'nearest_elementary_distance_km', 'nearest_middle_distance_km',
+       'nearest_high_distance_km', 'nearest_park_distance_km',
+       'nearest_park_area', 'num_subway_within_0_5', 'num_subway_within_1',
+       'num_subway_within_3', 'num_elementary_within_0_5',
+       'num_elementary_within_1', 'num_elementary_within_2',
+       'num_middle_within_0_5', 'num_middle_within_1', 'num_middle_within_2',
+       'num_high_within_0_5', 'num_high_within_1', 'num_high_within_2',
+       'num_park_within_0_8', 'num_park_within_1_5', 'num_park_within_2',
+       'area_floor_interaction']
```
Comment (sally-yeom) [inline modeling/CatBoost.py:52]: 해당 부분도 conifg로 관리하면 좋을 것 같아요.

File: modeling/CatBoost.py [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+
+
+# In[15]:
+
+
+holdout_start = 202307
+holdout_end = 202312
```
Comment (sally-yeom) [inline modeling/CatBoost.py:59]: 이 부분도 config로 관리 필요한 항목입니다.

File: modeling/CatBoost.py [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+    model = cb.CatBoostRegressor(**params)
+    
+    model.fit(X_train, y_train, eval_set=[(X_val, y_val)])
+    
+    holdout_pred = model.predict(X_holdout)
+    
+    holdout_mae = mean_absolute_error(y_holdout, holdout_pred)
+    holdout_rmse = root_mean_squared_error(y_holdout, holdout_pred)
+
+    trial.set_user_attr("rmse", holdout_rmse)
```
Comment (sally-yeom) [inline modeling/CatBoost.py:105]: 함수명이 적절하지 않습니다. python에서는 함수명을 Snake Case를 사용한 동사 형태의 이름으로 보통 설정합니다.

아래의 예시로 함수명을 참고해주세요.

```
def modify_list(my_list):
    my_list.append(4)
    print("Inside function:", my_list)
```

그리고 함수에 대한 주석 설명이 있으면 좋겠습니다.

File: modeling/FTTransformer(auto).py [hunk 15]
```diff
@@ -0,0 +1,364 @@
...
+# coding: utf-8
+
+# # 데이터 불러오기
+
+# In[2]:
+
+
+import pandas as pd
+import numpy as np
+from sklearn.metrics import root_mean_squared_error, mean_absolute_error
```
Comment (sally-yeom) [inline modeling/FTTransformer(auto).py:11]: 해당 파일도 마찬가지로 모든 import를 상단에 모아 위치시키면 좋겠습니다.

File: modeling/FTTransformer(auto).py [hunk 16]
```diff
@@ -0,0 +1,364 @@
...
+
+# # 하이퍼 파라미터 튜닝
+
+# In[ ]:
+
+
+import optuna
+from pytorch_tabular import TabularModel
+from pytorch_tabular.models import TabTransformerConfig
+from pytorch_tabular.config import DataConfig, OptimizerConfig, TrainerConfig
```
Comment (sally-yeom) [inline modeling/FTTransformer(auto).py:158]: 강제된 건 아니지만 보통 from ~~ import 문을 import 문 위에 위치시키는 경우가 많습니다.

File: modeling/FTTransformer(auto).py [hunk 17]
```diff
@@ -0,0 +1,364 @@
...
+    )
+    
+    # Set up trainer config
+    trainer_config = TrainerConfig(
+        batch_size=batch_size,
+        max_epochs=100,
+        early_stopping_mode="min",
+        early_stopping_patience=3,
+        min_epochs=1,
+    )
```
Comment (sally-yeom) [inline modeling/FTTransformer(auto).py:183]: 상수로 쓰이는 hyper params도 config로 관리되면 좋습니다.

File: modeling/Ensemble.py [hunk 18]
```diff
@@ -0,0 +1,630 @@
...
+# # 데이터 불러오기
+
+# In[1]:
+
+
+import pandas as pd
+import numpy as np
+from sklearn.metrics import root_mean_squared_error, mean_absolute_error
+from sklearn.model_selection import train_test_split
+import optuna
```
Comment (sally-yeom) [inline modeling/Ensemble.py:13]: 가독성을 위해 from ~~ import 구문, import 구문끼리 모아주시기 바랍니다.
현재는 두 구문이 섞여있습니다.

아래와 같이 되면 좋을 것 같아요.

```
from sklearn.metrics import root_mean_squared_error, mean_absolute_error
from sklearn.model_selection import train_test_split

import pandas as pd
import numpy as np
import optuna
```

File: modeling/Ensemble.py [hunk 19]
```diff
@@ -0,0 +1,630 @@
...
+
+RANDOM_SEED = 42
+np.random.seed(RANDOM_SEED)
+
+# 파일 경로 설정
+file_path = '../data/'
```
Comment (sally-yeom) [inline modeling/Ensemble.py:19]: 다른 파일에도 같은 코드가 있었습니다. 해당 부분은 config로 모아주고 공통 작업은 함수로 만들면 중복코드를 없앨 수 있습니다.

File: modeling/Ensemble.py [hunk 20]
```diff
@@ -0,0 +1,630 @@
...
+# In[18]:
+
+
+# 파일 불러오기
+interest_rate_df = pd.read_csv(file_path + 'interestRate.csv')
+
+interest_rate_df["contract_year_month"] = interest_rate_df["year_month"]
+interest_rate_df.drop("year_month", axis=1, inplace=True)
+
+interest_rate_df.loc[len(interest_rate_df)] = [3.56, 202406]
```
Comment (sally-yeom) [inline modeling/Ensemble.py:45]: [3.56, 202406] 와 같은 숫자가 코드에 direct로 들어가는 건 좋은 코드가 아닙니다.
해당 부분 상수로 지정해주시기 바랍니다. 아래 예시 참고해주세요.

```
INTEREST_CONST = [3.56, 202406] # 해당 Line은 config 파일로 분리
interest_rate_df.loc[len(interest_rate_df)] = INTEREST_CONST
```

File: modeling/Ensemble.py [hunk 21]
```diff
@@ -0,0 +1,630 @@
...
+print("Best trial:")
+print(f"MAE: {trial_lgb.value}")
+print(f"RMSE: {trial_lgb.user_attrs['rmse']}")
+print("Best hyperparameters: ", trial_lgb.params)
+
+
+# Best trial:   
+# MAE: 3433.937191036907   
+# RMSE: 5724.117930432466   
+# Best hyperparameters:  {'boosting_type': 'gbdt', 'learning_rate': 0.1686900811502798, 'n_estimators': 233, 'max_depth': 11, 'num_leaves': 84, 'subsample': 0.9116089240887149, 'colsample_bytree': 0.7268457056093285}
```
Comment (sally-yeom) [inline modeling/Ensemble.py:211]: 해당 파일에 적혀있는 모든 실험 메모는 코드에서 삭제해주셔야 합니다.
이 부분이 랩업레포트에 담기면 좋을 것 같아요.

File: modeling/Ensemble.py [hunk 22]
```diff
@@ -0,0 +1,630 @@
...
+best_params_xgb = {'device': 'cuda', 'objective': 'reg:absoluteerror', 'booster': 'gbtree', 'tree_method': 'hist', 'learning_rate': 0.1, 'max_depth': 9, 'max_leaves': 255, 'subsample': 0.9339458777384791, 'colsample_bytree': 0.789333445231045, 'eval_metric': 'mae', 'random_state': 42, 'num_boost_round': 306}
+best_params_xgb['n_estimators'] = best_params_xgb.pop('num_boost_round')
+best_params_cat =  {'loss_function': 'RMSE', 'iterations': 901, 'learning_rate': 0.05928029272464036, 'depth': 12, 'l2_leaf_reg': 3, 'eval_metric': 'MAE', 'random_seed': 42, 'logging_level': 'Silent', 'task_type': 'GPU'}
+lgb_model1 = lgb.LGBMRegressor(**best_params_lgb)
+xgb_model1 = xgb.XGBRegressor(**best_params_xgb)
+cat_model1 = cb.CatBoostRegressor(**best_params_cat)
+lgb_model1.fit(X_train_full, y_train_full)
+xgb_model1.fit(X_train_full, y_train_full)
+cat_model1.fit(X_train_full_cat, y_train_full_cat, cat_features=categorical_columns)
```
Comment (sally-yeom) [inline modeling/Ensemble.py:417]: 해당 코드들을 보니 lgb_model, xgb_model, cat_model에 대해 각각 3개의 모델을 생성하여 앙상블하고 있습니다.
hyper_params를 봤을 때 차이는 seed 값 뿐인 것 같습니다. 코드의 중복을 줄이고 가독성을 높이기 위해 seed를 인자로 넘기는 공통 함수를 만들어서 개발해주시기 바랍니다.

File: modeling/Stacking_Ensemble.py [hunk 23]
```diff
@@ -0,0 +1,232 @@
...
+    'learning_rate': 0.05,
+    'n_estimators': 500,
+    'max_depth': 15,
+    'num_leaves': 96,
+    'subsample': 0.9104369435751337,
+    'colsample_bytree': 0.5371328882033616,
+    'objective': 'regression_l1',
+    'random_state': 42,
+    'verbose': -1
+}
```
Comment (sally-yeom) [inline modeling/Stacking_Ensemble.py:128]: 실험을 통해 fix한 해당 params들도 config에 넣어주시면 좋습니다.

**Review by sally-yeom — COMMENTED** [review]
리뷰 제출 드립니다. 피드백 간단 요약 적어놓습니다.

 - 파일 변수명 형식 통일 필요
 - 주석 형식 통일 필요
 - config 파일 생성 필요
 - 함수 및 클래스화 생성 필요 (클래스까지는 아니더라도 중복 코드가 많이 보였으므로, 중복에 대해서 함수화는 필수로 보임)
        - 하지만 왠만하면 재사용성을 위해 클래스로 만드시는게 좋을 것 같습니다.

감사합니다:)
